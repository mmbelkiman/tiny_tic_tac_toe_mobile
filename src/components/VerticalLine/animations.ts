import {Animated} from 'react-native';

export const playOpacityAnimation = (animation: Animated.Value) => {
  Animated.timing(animation, {
    toValue: 0.6,
    duration: 90,
    useNativeDriver: false,
  }).start();
};

export const playWidthAnimation = (
  animation: Animated.Value,
  width: number,
) => {
  Animated.timing(animation, {
    toValue: width,
    duration: 80,
    useNativeDriver: false,
  }).start();
};

export const playShakeAnimation = (animation: Animated.Value) => {
  Animated.sequence([
    Animated.timing(animation, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }),
    Animated.timing(animation, {
      toValue: 0,
      duration: 200,
      useNativeDriver: false,
    }),
  ]).start();
};

export const getShakeStyle = (animation: Animated.Value) => ({
  transform: [
    {
      translateY: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [0, -5, 5, -5, 5, 0],
      }),
    },
  ],
});
