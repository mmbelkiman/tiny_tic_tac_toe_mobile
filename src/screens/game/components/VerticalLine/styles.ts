import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  line: {
    position: 'absolute',
    height: '90%',
    top: '5%',
    borderRadius: 10,
  },
});
