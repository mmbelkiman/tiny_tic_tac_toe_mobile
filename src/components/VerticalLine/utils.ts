import {PlayerMark} from '../../common/types';
import styles from './styles';
import {getShakeStyle} from './animations';
import {Animated} from 'react-native';
import {getPlayerColor} from '../../common/utils';

export const getLineStyle = (
  winner: PlayerMark,
  positionX: number,
  widthAnimation: Animated.Value,
  opacityAnimation: Animated.Value,
  shakeAnimation: Animated.Value,
) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    width: widthAnimation,
    opacity: opacityAnimation,
    left: positionX,
  },
  getShakeStyle(shakeAnimation),
];
