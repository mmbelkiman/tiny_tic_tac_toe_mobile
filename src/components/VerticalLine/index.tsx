import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {PlayerMark} from '../../common/types';
import {getLineStyle} from './utils';
import {playShakeAnimation, playWidthAnimation} from './animations';
import {playOpacityAnimation} from '../../common/animations';

interface VerticalLineProps {
  width: number;
  positionX: number;
  winner: PlayerMark;
  visible: boolean;
}

const VerticalLine: React.FC<VerticalLineProps> = ({
  width,
  positionX,
  winner,
  visible,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const widthAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (visible && winner) {
      playOpacityAnimation(opacityAnimation);
      playWidthAnimation(widthAnimation, width);
      playShakeAnimation(shakeAnimation);
    }
  }, [
    opacityAnimation,
    shakeAnimation,
    visible,
    width,
    widthAnimation,
    winner,
  ]);

  if (!visible || winner === null) {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle(
        winner,
        positionX,
        widthAnimation,
        opacityAnimation,
        shakeAnimation,
      )}
    />
  );
};

export default VerticalLine;
