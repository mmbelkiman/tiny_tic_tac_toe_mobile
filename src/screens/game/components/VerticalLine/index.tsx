import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {
  playOpacityAnimation,
  playShakeAnimation,
  playSizeAnimation,
} from '@game/common/animations';
import {getLineStyle} from './utils';
import {PlayerMark} from '@game/common/types';

interface VerticalLineProps {
  width: number;
  positionX: number;
  winner: PlayerMark;
}

const VerticalLine: React.FC<VerticalLineProps> = ({
  width,
  positionX,
  winner,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const widthAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    playOpacityAnimation(opacityAnimation);
    playSizeAnimation(widthAnimation, width);
    playShakeAnimation(shakeAnimation);
  }, [opacityAnimation, shakeAnimation, width, widthAnimation, winner]);

  if (winner === '') {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle(
        winner,
        positionX,
        widthAnimation,
        opacityAnimation,
        shakeAnimation,
      )}
    />
  );
};

export default VerticalLine;
