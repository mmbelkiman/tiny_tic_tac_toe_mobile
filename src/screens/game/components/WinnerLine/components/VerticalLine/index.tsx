import React from 'react';
import {Animated} from 'react-native';
import {getLineStyle} from './utils';

interface VerticalLineProps {
  width: number;
  positionX: number;
  color: string;
  widthAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
  testID?: string | number;
}

const VerticalLine: React.FC<VerticalLineProps> = ({
  positionX,
  color,
  widthAnimation,
  shakeAnimation,
  opacityAnimation,
  testID = '',
}) => {
  return (
    <Animated.View
      testID={`VerticalLine${testID}`}
      style={getLineStyle(
        color,
        positionX,
        widthAnimation,
        opacityAnimation,
        shakeAnimation,
      )}
    />
  );
};

export default VerticalLine;
