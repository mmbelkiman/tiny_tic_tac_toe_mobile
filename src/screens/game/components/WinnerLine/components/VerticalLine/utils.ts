import {Animated} from 'react-native';
import {getPlayerColor} from '../../../../common/utils';
import {PlayerMark} from '../../../../common/types';
import styles from './styles';

export const getShakeStyle = (animation: Animated.Value) => ({
  transform: [
    {
      translateY: animation.interpolate({
        inputRange: [0, 0.2, 0.4, 0.6, 0.8, 1],
        outputRange: [0, -5, 5, -5, 5, 0],
      }),
    },
  ],
});

export const getLineStyle = (
  winner: PlayerMark,
  positionX: number,
  widthAnimation: Animated.Value,
  opacityAnimation: Animated.Value,
  shakeAnimation: Animated.Value,
) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    width: widthAnimation,
    opacity: opacityAnimation,
    left: positionX,
  },
  getShakeStyle(shakeAnimation),
];
