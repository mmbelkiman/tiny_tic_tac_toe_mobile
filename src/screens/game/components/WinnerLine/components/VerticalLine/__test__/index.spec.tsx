import React from 'react';
import renderer from 'react-test-renderer';
import VerticalLine from '../index';
import {Animated} from 'react-native';

describe('VerticalLine Component', () => {
  it('should match the snapshot with a winner', () => {
    const widthAnimation = new Animated.Value(0);
    const opacityAnimation = new Animated.Value(0);
    const shakeAnimation = new Animated.Value(0);
    const tree = renderer
      .create(
        <VerticalLine
          width={100}
          positionX={50}
          color={'green'}
          widthAnimation={widthAnimation}
          opacityAnimation={opacityAnimation}
          shakeAnimation={shakeAnimation}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot without a winner', () => {
    const widthAnimation = new Animated.Value(0);
    const opacityAnimation = new Animated.Value(0);
    const shakeAnimation = new Animated.Value(0);

    const tree = renderer
      .create(
        <VerticalLine
          width={100}
          positionX={50}
          color={'blue'}
          shakeAnimation={shakeAnimation}
          opacityAnimation={opacityAnimation}
          widthAnimation={widthAnimation}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
