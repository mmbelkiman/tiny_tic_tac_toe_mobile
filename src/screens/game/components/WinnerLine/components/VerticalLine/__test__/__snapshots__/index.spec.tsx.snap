// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalLine Component should match the snapshot with a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "green",
      "borderRadius": 10,
      "height": "90%",
      "left": 50,
      "opacity": 0,
      "position": "absolute",
      "top": "5%",
      "transform": [
        {
          "translateY": 0,
        },
      ],
      "width": 0,
    }
  }
  testID="VerticalLine"
/>
`;

exports[`VerticalLine Component should match the snapshot without a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "blue",
      "borderRadius": 10,
      "height": "90%",
      "left": 50,
      "opacity": 0,
      "position": "absolute",
      "top": "5%",
      "transform": [
        {
          "translateY": 0,
        },
      ],
      "width": 0,
    }
  }
  testID="VerticalLine"
/>
`;
